pkg/antctl/commandoption.go

```diff
@@ -0,0 +1,350 @@
+	Agent bool
+	// Singleton should be true if the handler only returns a single object.
+	// The antctl assumes the response data as a slice of objects by default.
+	Singleton bool
+	// antctl will not parse the response from the server if this is true
+	PlainText bool
+	// The handler factory of the command.
+	HandlerFactory handlers.Factory
+	// ResponseStruct is the response object data struct of the command.
+	ResponseStruct interface{}
+	// AddonTransform is used to transform or fill the data received from the handler. It could be nil.
+	AddonTransform func(reader io.Reader, single bool) (interface{}, error)
+}
+
+// Validate checks if the CommandOption is valid.
+func (co *CommandOption) Validate() []error {
+	var errs []error
+	if len(co.Use) == 0 { // must have a use string
+		errs = append(errs, fmt.Errorf("the command does not have name"))
+	}
+	if co.HandlerFactory == nil { // must have a handler
+		errs = append(errs, fmt.Errorf("%s: no handler specified", co.Use))
+	}
+	if co.ResponseStruct == nil && !co.PlainText { // must define its response struct
+		errs = append(errs, fmt.Errorf("%s: no response struct", co.Use))
+	}
+	// Only one key arg is allowed.
+	var hasKey bool
+	for _, arg := range co.ArgOptions() {
+		if arg.Key && hasKey {
+			errs = append(errs, fmt.Errorf("%s: has more than one key field", co.Use))
+			break
+		} else if arg.Key {
+			hasKey = true
+		}
+	}
+	return errs
+}
+
+// ArgOptions returns the list of arguments that could be used in a CommandOption.
+func (co *CommandOption) ArgOptions() []*ArgOption {
+	var ret []*ArgOption
+	if co.ResponseStruct == nil {
+		return ret
+	}
+	t := reflect.Indirect(reflect.ValueOf(co.ResponseStruct)).Type()
+
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		argOpt := &ArgOption{FieldName: f.Name}
+
+		tags, err := structtag.Parse(string(f.Tag))
+		if err != nil { // Broken cli tags, skip this field
+			continue
+		}
+
+		jsonTag, err := tags.Get("json")
+		if err != nil {
+			argOpt.Name = strings.ToLower(f.Name)
+		} else {
+			argOpt.Name = jsonTag.Name
+		}
+
+		cliTag, err := tags.Get(TagKey)
+		if err == nil {
+			argOpt.Key = cliTag.Name == TagOptionKeyArg
+			argOpt.Usage = strings.Join(cliTag.Options, ", ")
+		}
+
+		ret = append(ret, argOpt)
+	}
+	return ret
+}
+
+func (co *CommandOption) decode(r io.Reader, single bool) (interface{}, error) {
+	var result interface{}
+
+	t := reflect.Indirect(reflect.ValueOf(co.ResponseStruct)).Type()
+
+	if single || co.Singleton {
+		ref := reflect.New(t)
+		err := json.NewDecoder(r).Decode(ref.Interface())
+		if err != nil {
+			return nil, err
+		}
+		result = ref.Interface()
+	} else {
+		ref := reflect.New(reflect.SliceOf(t))
+		err := json.NewDecoder(r).Decode(ref.Interface())
+		if err != nil {
+			return nil, err
+		}
+		result = reflect.Indirect(ref).Interface()
+	}
+
+	return result, nil
+}
+
+func (co *CommandOption) format(cmd *cobra.Command, r io.Reader, w io.Writer, single bool) (err error) {
```

Comment:
worth adding a comment here, for the sake of the developers who will have to work on this in the future. I suppose it is used to parse the reply from the server and output it to the user in the desired format?

Reply:
Added, pkg/antctl/commanddefinition.go:295